diff --git a/source/test/async_steps.js b/source/test/async_steps.js
index 6fcc77d..8b5bdd5 100644
--- a/source/test/async_steps.js
+++ b/source/test/async_steps.js
@@ -19,10 +19,8 @@ JS.Test.extend({
             block  = period;
             period = 'each';
           }
-          this.callSuper(function(resume) {
-            this.sync(function() {
-              this.exec(block, resume);
-            });
+          this.callSuper(period, function(resume) {
+            this.sync(function() { this.exec(block, resume) });
           });
         }
       });
@@ -41,12 +39,15 @@ JS.Test.extend({
         },
         
         __runNextStep__: function() {
-          var step = this.__stepQueue__.shift(), callback;
+          var step = this.__stepQueue__.shift();
           
           if (!step) {
             this.__runningSteps__ = false;
             if (!this.__stepCallbacks__) return;
-            while (callback = this.__stepCallbacks__.shift()) callback.call(this);
+            
+            for (i = 0, n = this.__stepCallbacks__.length; i < n; i++)
+              this.__stepCallbacks__.shift().call(this);
+            
             return;
           }
           
